core/kernel/lockdep.c

```diff
@@ -0,0 +1,397 @@
+#define LOCKDEP_NODE_PERM_MARK		BIT(1)
+/* Flag used during breadth-first search (print shortest cycle) */
+#define LOCKDEP_NODE_BFS_VISITED	BIT(2)
+
+/* Find node in graph or add it */
+static struct lockdep_node *lockdep_add_to_graph(
+				struct lockdep_node_head *graph,
+				uintptr_t lock_id)
+{
+	struct lockdep_node *node = NULL;
+
+	TAILQ_FOREACH(node, graph, link)
+		if (node->lock_id == lock_id)
+			return node;
+
+	node = calloc(1, sizeof(*node));
+	if (!node)
+		return NULL;
+
+	node->lock_id = lock_id;
+	STAILQ_INIT(&node->edges);
+	TAILQ_INSERT_TAIL(graph, node, link);
+
+	return node;
+}
+
+static vaddr_t *dup_call_stack(vaddr_t *stack)
+{
+	int n = 0;
+	vaddr_t *nstack = NULL;
+
+	if (!stack)
+		return NULL;
+	while (stack[n])
+		n++;
+	nstack = malloc(n * sizeof(vaddr_t));
+	if (!nstack)
+		return NULL;
+	memcpy(nstack, stack, n * sizeof(vaddr_t));
+	return nstack;
+}
+
+static void lockdep_print_call_stack(vaddr_t *stack)
+{
+	vaddr_t *p = NULL;
+
+	EMSG_RAW("Call stack:");
+	for (p = stack; p && *p; p++)
+		EMSG_RAW(" %#" PRIxPTR, *p);
+}
+
+static TEE_Result lockdep_add_edge(struct lockdep_node *from,
+				   struct lockdep_node *to,
+				   vaddr_t *call_stack_from,
+				   vaddr_t *call_stack_to,
+				   uintptr_t thread_id)
+{
+	struct lockdep_edge *edge = NULL;
+
+	STAILQ_FOREACH(edge, &from->edges, link)
+		if (edge->to == to)
+			return TEE_SUCCESS;
+
+	edge = calloc(1, sizeof(*edge));
+	if (!edge)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	edge->to = to;
+	edge->call_stack_from = dup_call_stack(call_stack_from);
+	edge->call_stack_to = dup_call_stack(call_stack_to);
+	edge->thread_id = thread_id;
+	STAILQ_INSERT_TAIL(&from->edges, edge, link);
+
+	return TEE_SUCCESS;
+}
+
+struct lockdep_bfs {
+	struct lockdep_node *node;
+	uintptr_t *path;
+	int pathlen;
+	TAILQ_ENTRY(lockdep_bfs) link;
+};
+
+TAILQ_HEAD(lockdep_bfs_head, lockdep_bfs);
+
+static void lockdep_bfs_queue_delete(struct lockdep_bfs_head *queue)
+{
+	struct lockdep_bfs *cur = NULL;
+	struct lockdep_bfs *next = NULL;
+
+	TAILQ_FOREACH_SAFE(cur, queue, link, next) {
+		TAILQ_REMOVE(queue, cur, link);
+		free(cur->path);
+		free(cur);
+	}
+}
+
+/*
+ * Print shortest cycle in @graph that contains @node.
+ * This function performs an iterative breadth-first search starting from @node,
+ * and stops when it reaches @node again. In each node we're tracking the path
+ * from the start node.
+ */
+static uintptr_t *lockdep_graph_get_shortest_cycle(struct lockdep_node *node)
+{
+	struct lockdep_bfs_head queue;
+	struct lockdep_bfs *qe = NULL;
+	uintptr_t *ret = NULL;
+
+	TAILQ_INIT(&queue);
+	node->flags |= LOCKDEP_NODE_BFS_VISITED;
+
+	qe = calloc(1, sizeof(*qe));
+	if (!qe)
+		goto out;
+	qe->node = node;
+	qe->path = malloc(sizeof(uintptr_t));
+	if (!qe->path)
+		goto out;
+	qe->path[0] = node->lock_id;
+	qe->pathlen = 1;
+	TAILQ_INSERT_TAIL(&queue, qe, link);
+
+	while (!TAILQ_EMPTY(&queue)) {
+		qe = TAILQ_FIRST(&queue);
+
+		struct lockdep_node *n = qe->node;
```

Comment:
minor: unused, remove.

Reply:
`n` unused.

Reply:
looks like you really don't want to get rid of this `n` :)

Reply:
It is used on line 151, actually ^_^

Reply:
oh.. yes. sorry.